crates/polars-utils/src/pl_str.rs

```diff
@@ -0,0 +1,124 @@
+use compact_str::CompactString;
+
+#[macro_export]
+macro_rules! format_pl_smallstr {
+    ($($arg:tt)*) => {{
+        use std::fmt::Write;
+
+        let mut string = PlSmallStr::const_default();
+        write!(string, $($arg)*).unwrap();
+        string
+    }}
+}
+
+/// String type that interns small strings and has O(1) clone.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct PlSmallStr(CompactString);
+
+impl PlSmallStr {
+    /// Initialize an empty string ""
+    #[inline(always)]
+    pub const fn const_default() -> Self {
+        Self(CompactString::const_new(""))
+    }
+
+    #[inline(always)]
+    pub const fn new_from_static(s: &'static str) -> Self {
+        Self(CompactString::const_new(s))
+    }
+
+    #[inline(always)]
+    pub fn new_from_str(s: &str) -> Self {
+        Self(CompactString::from(s))
+    }
+
+    #[inline(always)]
+    pub fn new_from_string(s: String) -> Self {
+        Self(CompactString::from(s))
+    }
+
+    #[inline(always)]
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+
+    #[inline(always)]
+    pub fn into_string(self) -> String {
+        self.0.into_string()
+    }
+}
+
+impl Default for PlSmallStr {
+    #[inline(always)]
+    fn default() -> Self {
+        Self(CompactString::const_new(""))
+    }
+}
+
+impl AsRef<str> for PlSmallStr {
+    #[inline(always)]
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl core::ops::Deref for PlSmallStr {
+    type Target = str;
+
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        self.0.as_str()
+    }
+}
+
+impl PartialEq<str> for PlSmallStr {
```

Comment:
I would make this implementation more generic, `PartialEq<T>` where `T: AsRef<str> + ?Sized`.

Then I would also add at the very least a `PartialEq` with `String` as well.

I also believe you have to do the inverse implementations, `PartialEq<PlSmallStr> for String` and such.